### PR TITLE
Single-server arangobackup CLI tool

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3884/

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3887/

File: arangosh/Backup/BackupFeature.cpp [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+  return {TRI_ERROR_NO_ERROR};
+}
+
+arangodb::Result getUptime(arangodb::httpclient::SimpleHttpClient& client, double& uptime) {
+  arangodb::Result result;
+
+  std::string const url = "/_admin/statistics";
+  std::unique_ptr<arangodb::httpclient::SimpleHttpResult> response(
+      client.request(arangodb::rest::RequestType::GET, url, nullptr, 0));
+  auto check = ::checkHttpResponse(client, response);
```
Comment (matthewvon) [inline arangosh/Backup/BackupFeature.cpp]: This is just a comment, not a fix request.  My understanding of Return Value Optimization (RVO) is that check represents a second potential Result variable used with "return".  That disables RVO.  If you had used result here, like every other return point, RVO could have applied.  Now this function does not need high performance, so this is just an observation.

File: arangosh/Backup/BackupFeature.cpp [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+
+void BackupFeature::validateOptions(std::shared_ptr<options::ProgramOptions> options) {
+  auto const& positionals = options->processingResult()._positionals;
+
+  if (1 == positionals.size()) {
+    _options.operation = positionals[0];
+  } else {
+    LOG_TOPIC(FATAL, Logger::BACKUP)
+        << "expected exactly one operation, got '"
+        << basics::StringUtils::join(positionals, ",") << "'";
```
Comment (matthewvon) [inline arangosh/Backup/BackupFeature.cpp]: "," intentional or ", " (with space) really the desired separator?

File: arangosh/Backup/BackupFeature.cpp [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+  auto const it = ::Operations.find(_options.operation);
+  if (it == ::Operations.end()) {
+    LOG_TOPIC(FATAL, Logger::BACKUP)
+        << "expected operation to be one of: " << operationList(", ");
+    FATAL_ERROR_EXIT();
+  }
+
+  if (_options.operation == ::OperationCreate) {
+    if (!_options.label.empty()) {
+      std::regex re = std::regex("^([a-zA-Z0-9\\-_]+)$", std::regex::ECMAScript);
```
Comment (matthewvon) [inline arangosh/Backup/BackupFeature.cpp]: Period is also allowed.  That is so release version numbers can be expressed.

File: arangosh/Backup/BackupFeature.cpp [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+  // extract vpack body from response
+  std::shared_ptr<VPackBuilder> parsedBody;
+  try {
+    parsedBody = response->getBodyVelocyPack();
+  } catch (...) {
+    result.reset(::ErrorMalformedJsonResponse);
+    return result;
+  }
+  VPackSlice const resBody = parsedBody->slice();
+  TRI_ASSERT(resBody.isObject());
```
Comment (dothebart) [inline arangosh/Backup/BackupFeature.cpp]: I don't think that asserts are the right things to do with a remote provided input data - they must not be trusted.

File: arangosh/Backup/BackupFeature.cpp [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+        double uptime = 0.0;
+        result = ::getUptime(*client, uptime);
+        if (result.ok() && uptime < originalUptime) {
+          LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+              << "...server back up and running!";
+          return result;
+        }
+        // server still down, or still up prior to restart; fall through, wait, retry
+      }
+    } catch (...) {
```
Comment (dothebart) [inline arangosh/Backup/BackupFeature.cpp]: is this really a good idea?

File: arangosh/Backup/BackupFeature.cpp [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+
+  if (backups.isEmptyArray()) {
+    LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+        << "There are no backups available.";
+  } else {
+    LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+        << "The following backups are available:";
+    for (auto const& backup : VPackArrayIterator(backups)) {
+      TRI_ASSERT(backup.isString());
+      LOG_TOPIC(INFO, arangodb::Logger::BACKUP) << " - '" << backup.copyString() << "'";
```
Comment (dothebart) [inline arangosh/Backup/BackupFeature.cpp]: is this properly cut'n'pasteable?

File: arangosh/Backup/BackupFeature.cpp [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+  std::shared_ptr<VPackBuilder> parsedBody;
+  try {
+    parsedBody = response->getBodyVelocyPack();
+  } catch (...) {
+    result.reset(::ErrorMalformedJsonResponse);
+    return result;
+  }
+  VPackSlice const resBody = parsedBody->slice();
+  VPackSlice const resultObject = resBody.get("result");
+  TRI_ASSERT(resultObject.isObject());
```
Comment (dothebart) [inline arangosh/Backup/BackupFeature.cpp]: don't lean on server replies to be correct.

File: arangosh/Backup/BackupFeature.cpp [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+    VPackSlice const previous = resultObject.get("previous");
+    TRI_ASSERT(previous.isString());
+
+    LOG_TOPIC(INFO, arangodb::Logger::BACKUP)
+        << "current state was saved as backup with identifier '"
+        << previous.copyString() << "'";
+  }
+
+  double originalUptime = 0.0;
+  if (options.maxWaitForRestart > 0.0) {
```
Comment (dothebart) [inline arangosh/Backup/BackupFeature.cpp]: do this before triggering the restore.

File: arangosh/Backup/BackupFeature.cpp (hunk 1 above)
Comment (danielhlarkin) [reply arangosh/Backup/BackupFeature.cpp]: Good catch.

File: arangosh/Backup/BackupFeature.cpp (hunk 2 above)
Comment (danielhlarkin) [reply arangosh/Backup/BackupFeature.cpp]: With space, good catch.

File: arangosh/Backup/BackupFeature.cpp (hunk 6 above)
Comment (danielhlarkin) [reply arangosh/Backup/BackupFeature.cpp]: You can copy the portion between the `'`s, yes. Should we omit the `'`s and just let it be whitespace around?

Comment (dothebart) [reply arangosh/Backup/BackupFeature.cpp]: idk whether other terminals don't do this right, xterm can, does putty?

Comment (danielhlarkin) [reply arangosh/Backup/BackupFeature.cpp]: I have no idea. There's an issue with copying and pasting strings in different terminals?

**Comment by danielhlarkin** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/3908/

**Pull request merged**